examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+        )
+
+        for batch_idx, batch in enumerate(iterable):
+            # end epoch if stopping training completely or max batches for this epoch reached
+            if self.should_stop or batch_idx >= limit_batches:
+                self.fabric.call("on_train_epoch_end")
+                return
+
+            self.fabric.call("on_train_batch_start", batch, batch_idx)
+
+            # check if optimizer should step in gradient accumulation
+            should_optim_step = self.global_step % self.grad_accum_steps == 0
+            if should_optim_step:
+                # currently only supports a single optimizer
+                self.fabric.call("on_before_optimizer_step", optimizer, 0)
+                # TODO: (lower priority) support optimizer without closure?
+
+                # optimizer step runs train step internally through closure
+                optimizer.step(
+                    partial(self.training_step, model=model, optimizer=optimizer, batch=batch, batch_idx=batch_idx)
+                )
+                self.fabric.call("on_before_zero_grad", optimizer)
+
+                optimizer.zero_grad()
+
+            else:
+                # gradient accumulation -> no optimizer step
+                self.training_step(model=model, optimizer=optimizer, batch=batch, batch_idx=batch_idx)
+
+            self.fabric.call("on_train_batch_end", self._current_train_return, batch, batch_idx)
+
+            # this guard ensures, we only step the scheduler once per global step
+            if should_optim_step:
+                self.step_scheduler(model, scheduler_cfg, level="step", current_value=self.global_step)
+
+            # add output values to progress bar
+            self._format_iterable(iterable, self._current_train_return, "train")
+
+            # only increase global step if optimizer stepped
+            self.global_step += int(should_optim_step)
+
+            # stopping criterion on step level
+            if self.max_steps is not None and self.global_step >= self.max_steps:
+                self.should_stop = True
+                break
+
+        self.fabric.call("on_train_epoch_end")
+
+    def val_loop(
+        self, model: LightningModule, val_loader: Optional[DataLoader], limit_batches: Union[int, float] = float("inf")
+    ):
+        # no validation if val_loader wasn't passed
+        if val_loader is None:
+            return
+
+        # no validation but warning if val_loader was passed, but validation_step not implemented
+        elif val_loader is not None and not is_overridden("validation_step", _unwrap_objects(model), LightningModule):
+            rank_zero_warn(
+                "Your LightningModule does not have a validation_step implemented, "
+                "but you passed a validation dataloder. Skipping Validation."
+            )
+            return
+
+        self.fabric.call("on_validation_model_eval")  # calls `model.eval()`
+
+        # TODO: inference mode?
```

Comment:
opinions on no-grad vs inference mode? In validation I'd do no-grad and only in prediction and testing I'd do inference mode, but also open for changes here :)